bundles/admin/admin-layereditor/view/AdminLayerForm/VisualizationTabPane/DynamicScreensSpaceErrorOptions.jsx

```diff
@@ -0,0 +1,157 @@
+        } />
+    </StyledDynamicScreenSpaceErrorOption>
+};
+
+const DynamicScreenSpaceErrorHeightFalloff = ({ layer, controller, defaultValue }) => {
+    const stepValue = 0.01;
+    const value = layer.options.dynamicScreenSpaceErrorHeightFalloff === undefined ? defaultValue : layer.options.dynamicScreenSpaceErrorHeightFalloff;
+    return <StyledDynamicScreenSpaceErrorOption>
+        <Message messageKey='dynamicScreenSpaceErrorHeightFalloff' />
+        <StyledInputContainer>
+            <NumberInput
+                defaultValue={defaultValue}
+                step={stepValue}
+                value={value}
+                onChange={eventValue => controller.setOptions({ ...layer.options, dynamicScreenSpaceErrorHeightFalloff: eventValue })}
+            />
+        </StyledInputContainer>
+        <InfoTooltip message={
+            <Fragment>
+                <pre>{dynamicScreenSpaceErrorHeightFalloffTemplate}</pre>
+            </Fragment>
+        } />
+    </StyledDynamicScreenSpaceErrorOption>
+};
+
+export const DynamicScreensPaceErrorOptions = ({ layer, controller }) => {
+
+    const defaultDynamicScreenSpaceError = false;
+    const defaultDynamicScreenSpaceErrorDensity = 0.00278;
+    const defaultDynamicScreenSpaceErrorFactor = 4.0;
+    const defaultDynamicScreenSpaceErrorHeightFalloff = 0.25;    
+
+    // If layer options contain default values, remove them.
+    layer.options.hasOwnProperty('dynamicScreenSpaceError') && layer.options.dynamicScreenSpaceError === defaultDynamicScreenSpaceError && delete layer.options['dynamicScreenSpaceError'];
+    layer.options.hasOwnProperty('dynamicScreenSpaceErrorDensity') && layer.options.dynamicScreenSpaceErrorDensity === defaultDynamicScreenSpaceErrorDensity && delete layer.options['dynamicScreenSpaceErrorDensity'];
+    layer.options.hasOwnProperty('dynamicScreenSpaceErrorFactor') && layer.options.dynamicScreenSpaceErrorFactor === defaultDynamicScreenSpaceErrorFactor && delete layer.options['dynamicScreenSpaceErrorFactor'];
+    layer.options.hasOwnProperty('dynamicScreenSpaceErrorHeightFalloff') && layer.options.dynamicScreenSpaceErrorHeightFalloff === defaultDynamicScreenSpaceErrorHeightFalloff && delete layer.options['dynamicScreenSpaceErrorHeightFalloff'];
+
+    return <StyledFormField>
+        <StyledDynamicScreenSpaceErrorOptionsContainer>
+            <DynamicScreenSpaceError layer={layer} controller={controller} defaultValue={defaultDynamicScreenSpaceError}/>
+            <DynamicScreenSpaceErrorDensity layer={layer} controller={controller} defaultValue={defaultDynamicScreenSpaceErrorDensity}/>
+            <DynamicScreenSpaceErrorFactor layer={layer} controller={controller} defaultValue={defaultDynamicScreenSpaceErrorFactor}/>
+            <DynamicScreenSpaceErrorHeightFalloff layer={layer} controller={controller} defaultValue={defaultDynamicScreenSpaceErrorHeightFalloff}/>
+        </StyledDynamicScreenSpaceErrorOptionsContainer>
+    </StyledFormField>
```

Comment:
This approach only removes the value for rendering, not saving. How about something like this:
```suggestion
    const defaults = {
        dynamicScreenSpaceError: false,
        dynamicScreenSpaceErrorDensity: 0.00278,
        dynamicScreenSpaceErrorFactor: 4.0,
        dynamicScreenSpaceErrorHeightFalloff: 0.25
    };

    const getDefaultRemovingController = (key) => {
        return {
            setOptions = (opts) => {
                if (defaults[key] === opts[key]) {
                    // setting default value -> remove it
                    delete opts[key];
                }
                controller.setOptions(opts);
            }
        };
    };
    const getAttributes = (fieldName) => {
        return {
            layer,
            controller: getDefaultRemovingController(fieldName),
            defaultValue: defaults[fieldName]
        };
    }

    return <StyledFormField>
        <StyledDynamicScreenSpaceErrorOptionsContainer>
            <DynamicScreenSpaceError ...getAttributes('dynamicScreenSpaceError') />
            <DynamicScreenSpaceErrorDensity  ...getAttributes('dynamicScreenSpaceErrorDensity') />
            <DynamicScreenSpaceErrorFactor  ...getAttributes('dynamicScreenSpaceErrorFactor') />
            <DynamicScreenSpaceErrorHeightFalloff  ...getAttributes('dynamicScreenSpaceErrorHeightFalloff') />
        </StyledDynamicScreenSpaceErrorOptionsContainer>
    </StyledFormField>
```
Though I would refactor the child components to have a onChange() callback that could be used to call the setOptions() of the actual controller without needing to fake it, but this was easier to link as an example. Also the defaults could be moved out of the component to a constant variable in the file.

Reply:
Just tested this again and seems to work and options are saved to database if they are not default values and save button is pressed.
setOptions() is called from each component when inputs are changed.
Layer options is compared to react component state on each render and removed if layer option matches default value.

https://github.com/oskariorg/oskari-frontend/pull/1397/files/7c1105ccbfd8397ffbea344179fd4241898c7a6f#diff-222746e53136bc07b49bd3b0ba18f6dc1c4c4078cedaad4e3d78d99f731a991bR46-R130

Or did I miss something?

Reply:
I think it makes more sense to _not_ set the value and remove the value when it it's changed to the default one _when it's being modified_  since we don't want to save it. Your solution always sets the value and then tries to correct the data on render assuming it is re-rendered and that the state is mutable. The component might not be rendered again due to logic change or js error. That's the difference on our approaches: the proposed one resets the value in the state when it's changed to a default value, your approach writes an unwanted state and removes the value when the component is rendered. Changing state while rendering will lead to non-trivial bugs later on.

In practice there's currently little difference to either one but if we ever move to immutable state this won't work. Also if we update the code to save the changes without the user clicking the save button the approach where data is changed while being rendered doesn't work and it won't be obvious why the saved layer isn't the same the user sees on the screen.

Reply:
Oh yes I get your point and your approach is probably more suitable.
Unnecessary re-renders are indeed not very good to have.
I will do the changes like you proposed.